lib/pos_blockchain.go

```diff
@@ -57,66 +51,179 @@ func (bc *Blockchain) processBlockPoS(block *MsgDeSoBlock, currentView uint64, v
+// Base case: If block is VALIDATED or VALIDATE_FAILED, return the BlockNode as-is.
+// Recursive case: If the block is not VALIDATED or VALIDATE_FAILED in the blockIndex, we will perform all validations
+// and add the block to the block index with the appropriate status (VALIDATED OR VALIDATE_FAILED) and return the new BlockNode.
+// Edge case: Something goes wrong that doesn't result in the block being marked VALIDATE or VALIDATE_FAILED. In this
+// case, we will add the block to the block index with status STORED and return the BlockNode.
+func (bc *Blockchain) isValidPoSBlock(block *MsgDeSoBlock) (*BlockNode, error) {
+	var storeInBlockIndexErr error
+	blockHash, err := block.Header.Hash()
+	if err != nil {
+		return nil, errors.Wrapf(err, "isValidPoSBlock: Problem hashing block %v", block)
+	}
+	// Base case - Check if the block itself is validated or validate failed. If so, we can return early.
+	blockNode, exists := bc.blockIndex[*blockHash]
+	if exists {
+		if blockNode.IsValidateFailed() {
+			return blockNode, errors.New("isValidPoSBlock: Block already marked as ValidateFailed")
+		}
+		if blockNode.IsValidated() {
+			return blockNode, nil
+		}
+	}
 	// Make sure its parent is validated. If not, we need to validate it first by calling processBlockPoS.
 	// Note that if any other ancestor is not validated, calling processBlockPoS on the parent will
 	// end up calling processBlockPoS on all of its ancestors that aren't validated yet.
 	parentBlockNode, exists := bc.blockIndex[*block.Header.PrevBlockHash]
 	if !exists {
-		return false, false, nil, errors.New("processBlockPoS: Parent block not found in block index. This should never happen.")
+		// We should never really hit this, but if we do, we upsert the block w/ StatusBlockStored
+		// but not StatusBlockValidateFailed. This makes sure this function is idempotent.
+		blockNode, err = bc.storeBlockInBlockIndex(block)
+		if err != nil {
+			return blockNode, errors.Wrap(err, "isValidPoSBlock: Problem adding block to block index: ")
+		}
+		return blockNode, errors.New("isValidPoSBlock: Parent block not found in block index. This should never happen.")
 	}
 	if parentBlockNode.IsValidateFailed() {
 		// We'll never add this to the best chain because its parent is invalid.
 		// If its parent is ValidateFailed, we can update the block index with the block status ValidateFailed
-		if err = bc.storeValidateFailedBlockInBlockIndex(block); err != nil {
-			return false, false, nil, errors.Wrap(err, "processBlockPoS: Problem adding validate failed block to block index")
+		blockNode, err = bc.storeValidateFailedBlockInBlockIndex(block)
+		if err != nil {
+			return blockNode, errors.Wrap(err, "isValidPoSBlock: Problem adding validate failed block to block index")
 		}
-		return false, false, nil, errors.New("processBlockPoS: block's parent failed validation: ")
+		return blockNode, errors.New("isValidPoSBlock: block's parent failed validation: ")
 	}
 	if !parentBlockNode.IsValidated() {
```

Comment:
IMO, you can delete this block to simplify the nested checks below. Rather than checking the parentBlockNode's status manually in the blockIndex, you can fetch the parent block from DB and call `isValidBlock(parentBlock)` on it directly. This eliminates the need for checking the parent block's status twice.

Granted, it is less efficient than simply performing an in-memory check of the block index first, but Badger is quite good at caching recently used data in memory that I'm sure in practice, the performance difference is negligible. I'd bias towards starting with the more concise (less nested) solution and optimizing later if needed.